File: backend/src/kb/localStore.ts
import { mkdir, readFile, writeFile } from "fs/promises";
import { dirname, join } from "path";
import { ObjectId } from "mongodb";

type StoredDoc = Record<string, any>;
type StoredData = {
  kb_chunks: StoredDoc[];
  kb_files: StoredDoc[];
};
type SortSpec = Record<string, 1 | -1>;
type ProjectionSpec = Record<string, 0 | 1>;

const LOCAL_STORE_PATH = join(process.cwd(), "data", "kb-local-store.json");
const EMPTY_DATA: StoredData = {
  kb_chunks: [],
  kb_files: [],
};

let dataPromise: Promise<StoredData> | null = null;
let writeQueue: Promise<void> = Promise.resolve();

function cloneValue<T>(value: T): T {
  if (value instanceof ObjectId) {
    return new ObjectId(value.toHexString()) as T;
  }
  if (value instanceof Date) {
    return new Date(value.getTime()) as T;
  }
  if (Array.isArray(value)) {
    return value.map((item) => cloneValue(item)) as T;
  }
  if (value && typeof value === "object") {
    return Object.fromEntries(
      Object.entries(value).map(([key, item]) => [key, cloneValue(item)])
    ) as T;
  }
  return value;
}

function cloneDoc<T>(doc: T): T {
  return cloneValue(doc);
}

function objectIdToString(value: unknown): string | null {
  if (value instanceof ObjectId) return value.toHexString();
  if (value && typeof (value as any).toHexString === "function") {
    return (value as any).toHexString();
  }
  return typeof value === "string" ? value : null;
}

function normalizeDocForMemory(doc: StoredDoc): StoredDoc {
  const next = { ...doc };
  const id = objectIdToString(next._id);
  if (id && ObjectId.isValid(id)) {
    next._id = new ObjectId(id);
  }
  if (typeof next.uploadedAt === "string") {
    next.uploadedAt = new Date(next.uploadedAt);
  }
  return next;
}

function normalizeDocForDisk(doc: StoredDoc): StoredDoc {
  const next = cloneDoc(doc);
  const id = objectIdToString(next._id);
  if (id) next._id = id;
  return next;
}

async function loadData(): Promise<StoredData> {
  if (!dataPromise) {
    dataPromise = (async () => {
      try {
        const raw = await readFile(LOCAL_STORE_PATH, "utf8");
        const parsed = JSON.parse(raw) as Partial<StoredData>;
        return {
          kb_chunks: Array.isArray(parsed.kb_chunks)
            ? parsed.kb_chunks.map(normalizeDocForMemory)
            : [],
          kb_files: Array.isArray(parsed.kb_files)
            ? parsed.kb_files.map(normalizeDocForMemory)
            : [],
        };
      } catch (error: any) {
        if (error?.code === "ENOENT") {
          return cloneDoc(EMPTY_DATA);
        }
        throw error;
      }
    })();
  }

  return dataPromise;
}

async function persistData(data: StoredData): Promise<void> {
  const write = async () => {
    await mkdir(dirname(LOCAL_STORE_PATH), { recursive: true });
    const diskData: StoredData = {
      kb_chunks: data.kb_chunks.map(normalizeDocForDisk),
      kb_files: data.kb_files.map(normalizeDocForDisk),
    };
    await writeFile(LOCAL_STORE_PATH, JSON.stringify(diskData, null, 2));
  };

  writeQueue = writeQueue.then(write, write);
  return writeQueue;
}

function getPathValue(doc: StoredDoc, path: string): unknown {
  return path.split(".").reduce<unknown>((value, key) => {
    if (value && typeof value === "object") {
      return (value as any)[key];
    }
    return undefined;
  }, doc);
}

function setPathValue(doc: StoredDoc, path: string, value: unknown): void {
  const keys = path.split(".");
  let current = doc;
  for (const key of keys.slice(0, -1)) {
    if (!current[key] || typeof current[key] !== "object") {
      current[key] = {};
    }
    current = current[key];
  }
  current[keys[keys.length - 1]] = value;
}

function valuesEqual(left: unknown, right: unknown): boolean {
  const leftId = objectIdToString(left);
  const rightId = objectIdToString(right);
  if (leftId || rightId) return leftId === rightId;
  return left === right;
}

function matchesType(value: unknown, type: string): boolean {
  if (type === "array") return Array.isArray(value);
  if (type === "string") return typeof value === "string";
  if (type === "number") return typeof value === "number";
  if (type === "date") return value instanceof Date;
  if (type === "object") {
    return !!value && typeof value === "object" && !Array.isArray(value);
  }
  return false;
}

function matchesCondition(value: unknown, condition: unknown): boolean {
  if (
    condition &&
    typeof condition === "object" &&
    !(condition instanceof Date) &&
    !(condition instanceof ObjectId) &&
    !Array.isArray(condition)
  ) {
    for (const [operator, expected] of Object.entries(condition)) {
      if (operator === "$type") {
        if (!matchesType(value, String(expected))) return false;
      } else if (operator === "$exists") {
        if ((value !== undefined) !== Boolean(expected)) return false;
      } else if (operator === "$ne") {
        if (valuesEqual(value, expected)) return false;
      } else {
        return false;
      }
    }
    return true;
  }

  return valuesEqual(value, condition);
}

function matchesFilter(doc: StoredDoc, filter: StoredDoc = {}): boolean {
  for (const [key, condition] of Object.entries(filter)) {
    if (key === "$or") {
      if (
        !Array.isArray(condition) ||
        !condition.some((item) => matchesFilter(doc, item))
      ) {
        return false;
      }
      continue;
    }

    if (!matchesCondition(getPathValue(doc, key), condition)) {
      return false;
    }
  }
  return true;
}

function applyProjection(doc: StoredDoc, projection?: ProjectionSpec): StoredDoc {
  if (!projection) return cloneDoc(doc);

  const includeKeys = Object.entries(projection)
    .filter(([_, value]) => value === 1)
    .map(([key]) => key);
  if (includeKeys.length > 0) {
    const next: StoredDoc = {};
    for (const key of includeKeys) {
      const value = getPathValue(doc, key);
      if (value !== undefined) setPathValue(next, key, value);
    }
    if (projection._id !== 0 && doc._id !== undefined) {
      next._id = doc._id;
    }
    return cloneDoc(next);
  }

  const next = cloneDoc(doc);
  for (const [key, value] of Object.entries(projection)) {
    if (value === 0) delete next[key];
  }
  return next;
}

function cosineSimilarity(left: number[], right: number[]): number {
  if (!left.length || left.length !== right.length) return 0;

  let dot = 0;
  let leftNorm = 0;
  let rightNorm = 0;
  for (let index = 0; index < left.length; index++) {
    dot += left[index] * right[index];
    leftNorm += left[index] * left[index];
    rightNorm += right[index] * right[index];
  }
  if (!leftNorm || !rightNorm) return 0;
  return dot / (Math.sqrt(leftNorm) * Math.sqrt(rightNorm));
}

class LocalCursor {
  private docsPromise: Promise<StoredDoc[]>;

  constructor(docsPromise: Promise<StoredDoc[]>) {
    this.docsPromise = docsPromise;
  }

  sort(spec: SortSpec): LocalCursor {
    this.docsPromise = this.docsPromise.then((docs) => {
      const entries = Object.entries(spec);
      return [...docs].sort((left, right) => {
        for (const [key, direction] of entries) {
          const leftValue = getPathValue(left, key) as any;
          const rightValue = getPathValue(right, key) as any;
          if (leftValue === rightValue) continue;
          if (leftValue === undefined) return 1;
          if (rightValue === undefined) return -1;
          return leftValue > rightValue ? direction : -direction;
        }
        return 0;
      });
    });
    return this;
  }

  limit(count: number): LocalCursor {
    this.docsPromise = this.docsPromise.then((docs) => docs.slice(0, count));
    return this;
  }

  async toArray(): Promise<StoredDoc[]> {
    return (await this.docsPromise).map(cloneDoc);
  }
}

class LocalDb {
  async command(command: StoredDoc): Promise<StoredDoc> {
    if (command.ping === 1) return { ok: 1, local: true };
    return { ok: 1, local: true };
  }
}

export class LocalKbCollection {
  readonly db = new LocalDb();
  readonly isLocalKbCollection = true;
  private name: keyof StoredData;

  constructor(name: keyof StoredData) {
    this.name = name;
  }

  async createIndex(_keys: StoredDoc, options?: { name?: string }): Promise<string> {
    return options?.name ?? "local_index";
  }

  find(filter: StoredDoc = {}, options?: { projection?: ProjectionSpec }): LocalCursor {
    const docsPromise = loadData().then((data) =>
      data[this.name]
        .filter((doc) => matchesFilter(doc, filter))
        .map((doc) => applyProjection(doc, options?.projection))
    );
    return new LocalCursor(docsPromise);
  }

  async findOne(
    filter: StoredDoc = {},
    options?: { projection?: ProjectionSpec }
  ): Promise<StoredDoc | null> {
    const docs = await this.find(filter, options).limit(1).toArray();
    return docs[0] ?? null;
  }

  async distinct(field: string): Promise<unknown[]> {
    const data = await loadData();
    const values = data[this.name]
      .map((doc) => getPathValue(doc, field))
      .filter((value) => value !== undefined);
    return Array.from(new Set(values));
  }

  async insertOne(doc: StoredDoc): Promise<{ acknowledged: true; insertedId: unknown }> {
    const data = await loadData();
    const next = normalizeDocForMemory({
      ...doc,
      _id: doc._id ?? new ObjectId(),
    });
    data[this.name].push(next);
    await persistData(data);
    return { acknowledged: true, insertedId: next._id };
  }

  async insertMany(
    docs: StoredDoc[]
  ): Promise<{ acknowledged: true; insertedCount: number }> {
    const data = await loadData();
    data[this.name].push(
      ...docs.map((doc) =>
        normalizeDocForMemory({
          ...doc,
          _id: doc._id ?? new ObjectId(),
        })
      )
    );
    await persistData(data);
    return { acknowledged: true, insertedCount: docs.length };
  }

  async updateOne(
    filter: StoredDoc,
    update: { $set?: StoredDoc }
  ): Promise<{ acknowledged: true; matchedCount: number; modifiedCount: number }> {
    const data = await loadData();
    const doc = data[this.name].find((item) => matchesFilter(item, filter));
    if (!doc) {
      return { acknowledged: true, matchedCount: 0, modifiedCount: 0 };
    }
    if (update.$set) Object.assign(doc, update.$set);
    await persistData(data);
    return { acknowledged: true, matchedCount: 1, modifiedCount: 1 };
  }

  async deleteOne(filter: StoredDoc): Promise<{ acknowledged: true; deletedCount: number }> {
    const data = await loadData();
    const index = data[this.name].findIndex((doc) => matchesFilter(doc, filter));
    if (index === -1) {
      return { acknowledged: true, deletedCount: 0 };
    }
    data[this.name].splice(index, 1);
    await persistData(data);
    return { acknowledged: true, deletedCount: 1 };
  }

  async deleteMany(filter: StoredDoc): Promise<{ acknowledged: true; deletedCount: number }> {
    const data = await loadData();
    const before = data[this.name].length;
    data[this.name] = data[this.name].filter((doc) => !matchesFilter(doc, filter));
    const deletedCount = before - data[this.name].length;
    if (deletedCount > 0) await persistData(data);
    return { acknowledged: true, deletedCount };
  }

  async countDocuments(filter: StoredDoc = {}): Promise<number> {
    const data = await loadData();
    return data[this.name].filter((doc) => matchesFilter(doc, filter)).length;
  }

  aggregate(pipeline: StoredDoc[]): LocalCursor {
    const docsPromise = loadData().then((data) => {
      const vectorSearch = pipeline.find((stage) => stage.$vectorSearch)
        ?.$vectorSearch;
      if (!vectorSearch) return [];

      const queryVector = Array.isArray(vectorSearch.queryVector)
        ? vectorSearch.queryVector
        : [];
      const limit =
        typeof vectorSearch.limit === "number" ? vectorSearch.limit : 5;

      return data[this.name]
        .filter((doc) => typeof doc.text === "string" && Array.isArray(doc.embedding))
        .map((doc) => {
          const score = cosineSimilarity(queryVector, doc.embedding);
          return {
            ...doc,
            score: (Math.max(-1, Math.min(1, score)) + 1) / 2,
          };
        })
        .sort((left, right) => right.score - left.score)
        .slice(0, limit)
        .map((doc) => {
          const project = pipeline.find((stage) => stage.$project)?.$project;
          return applyProjection(doc, project);
        });
    });
    return new LocalCursor(docsPromise);
  }
}

const localCollections = {
  kb_chunks: new LocalKbCollection("kb_chunks"),
  kb_files: new LocalKbCollection("kb_files"),
};

export function getLocalKbCollection(name: keyof StoredData): LocalKbCollection {
  return localCollections[name];
}

export function isLocalKbCollection(collection: unknown): boolean {
  return !!(
    collection &&
    typeof collection === "object" &&
    (collection as any).isLocalKbCollection
  );
}

File: backend-py/src/kb/local_store.py
"""本地 JSON 文件存储 —— 模拟 MongoDB Collection API。"""

import json
import logging
import math
import os
from bson import ObjectId
from copy import deepcopy

logger = logging.getLogger(__name__)

LOCAL_STORE_PATH = os.path.join(os.getcwd(), "data", "kb-local-store.json")
EMPTY_DATA = {"kb_chunks": [], "kb_files": []}

_data_promise = None


def _load_data():
    """加载本地存储数据。"""
    global _data_promise
    if _data_promise is not None:
        return _data_promise

    try:
        if os.path.exists(LOCAL_STORE_PATH):
            with open(LOCAL_STORE_PATH, "r", encoding="utf-8") as f:
                parsed = json.load(f)
        else:
            parsed = deepcopy(EMPTY_DATA)
    except Exception:
        logger.warning("[localStore] Failed to load, using empty data")
        parsed = deepcopy(EMPTY_DATA)

    # 反序列化 ObjectId
    for key in ("kb_chunks", "kb_files"):
        for doc in parsed.get(key, []):
            if isinstance(doc.get("_id"), str) and ObjectId.is_valid(doc["_id"]):
                doc["_id"] = ObjectId(doc["_id"])

    _data_promise = parsed
    return parsed


def _persist_data(data):
    """持久化数据到 JSON 文件。"""
    os.makedirs(os.path.dirname(LOCAL_STORE_PATH), exist_ok=True)
    disk_data = deepcopy(data)
    for key in ("kb_chunks", "kb_files"):
        for doc in disk_data.get(key, []):
            if isinstance(doc.get("_id"), ObjectId):
                doc["_id"] = str(doc["_id"])

    with open(LOCAL_STORE_PATH, "w", encoding="utf-8") as f:
        json.dump(disk_data, f, ensure_ascii=False, indent=2, default=str)


def _get_path_value(doc, path):
    """获取嵌套字段值，如 'metadata.source'。"""
    for key in path.split("."):
        if isinstance(doc, dict):
            doc = doc.get(key)
        else:
            return None
    return doc


def _set_path_value(doc, path, value):
    """设置嵌套字段值。"""
    keys = path.split(".")
    for key in keys[:-1]:
        if key not in doc or not isinstance(doc[key], dict):
            doc[key] = {}
        doc = doc[key]
    doc[keys[-1]] = value


def _values_equal(left, right):
    """比较两个值是否相等，兼容 ObjectId。"""
    if isinstance(left, ObjectId):
        return str(left) == str(right) if isinstance(right, (ObjectId, str)) else False
    if isinstance(right, ObjectId):
        return str(left) == str(right) if isinstance(left, (ObjectId, str)) else False
    return left == right


def _matches_type(value, type_str):
    """检查值是否匹配指定 BSON 类型。"""
    if type_str == "array":
        return isinstance(value, list)
    if type_str == "string":
        return isinstance(value, str)
    if type_str == "number":
        return isinstance(value, (int, float)) and not isinstance(value, bool)
    if type_str == "object":
        return isinstance(value, dict) and not isinstance(value, list)
    return False


def _matches_condition(value, condition):
    """检查值是否匹配条件。"""
    if isinstance(condition, dict):
        for operator, expected in condition.items():
            if operator == "$type":
                if not _matches_type(value, str(expected)):
                    return False
            elif operator == "$exists":
                exists = value is not None
                if exists != bool(expected):
                    return False
            elif operator == "$ne":
                if _values_equal(value, expected):
                    return False
            else:
                return False
        return True

    return _values_equal(value, condition)


def _matches_filter(doc, filter_dict=None):
    """检查文档是否匹配过滤条件。"""
    if not filter_dict:
        return True

    for key, condition in filter_dict.items():
        if key == "$or":
            if not isinstance(condition, list):
                return False
            if not any(_matches_filter(doc, item) for item in condition):
                return False
            continue

        if not _matches_condition(_get_path_value(doc, key), condition):
            return False

    return True


def _apply_projection(doc, projection=None):
    """应用字段投影。"""
    if not projection:
        return deepcopy(doc)

    include_keys = [k for k, v in projection.items() if v == 1]
    if include_keys:
        result = {}
        for key in include_keys:
            val = _get_path_value(doc, key)
            if val is not None:
                _set_path_value(result, key, val)
        if projection.get("_id") != 0 and "_id" in doc:
            result["_id"] = doc["_id"]
        return result

    result = deepcopy(doc)
    for key, v in projection.items():
        if v == 0:
            result.pop(key, None)
    return result


def _cosine_similarity(a, b):
    """计算余弦相似度。"""
    if not a or not b or len(a) != len(b):
        return 0.0
    dot = sum(x * y for x, y in zip(a, b))
    norm_a = math.sqrt(sum(x * x for x in a))
    norm_b = math.sqrt(sum(x * x for x in b))
    if not norm_a or not norm_b:
        return 0.0
    return dot / (norm_a * norm_b)


class LocalCursor:
    """模拟 MongoDB Cursor 的链式调用。"""

    def __init__(self, docs):
        self._docs = list(docs) if docs else []

    def sort(self, spec):
        """排序，spec 如 {'uploadedAt': -1}。"""
        items = list(self._docs)
        entries = list(spec.items()) if isinstance(spec, dict) else spec

        for key, direction in reversed(entries):
            items.sort(
                key=lambda d: _get_path_value(d, key) if _get_path_value(d, key) is not None else "",
                reverse=(direction == -1),
            )
        self._docs = items
        return self

    def limit(self, n):
        """截取前 n 条。"""
        self._docs = self._docs[:n]
        return self

    def to_list(self, length=None):
        """转换为列表。"""
        result = deepcopy(self._docs)
        if length is not None:
            result = result[:length]
        return result


class LocalKbCollection:
    """本地 JSON 文件存储，模拟 MongoDB Collection API。"""

    isLocalKbCollection = True

    def __init__(self, name):
        self._name = name  # "kb_chunks" or "kb_files"

    # --- 索引 ---
    async def create_index(self, keys, **kwargs):
        return kwargs.get("name", "local_index")

    # --- 查询 ---
    def find(self, filter_dict=None, projection=None):
        data = _load_data()
        docs = [
            _apply_projection(doc, projection)
            for doc in data[self._name]
            if _matches_filter(doc, filter_dict)
        ]
        return LocalCursor(docs)

    async def find_one(self, filter_dict=None, projection=None):
        docs = self.find(filter_dict, projection).limit(1).to_list()
        return docs[0] if docs else None

    # --- 写入 ---
    async def insert_one(self, doc):
        data = _load_data()
        doc = deepcopy(doc)
        if "_id" not in doc:
            doc["_id"] = ObjectId()
        data[self._name].append(doc)
        _persist_data(data)

        class InsertOneResult:
            def __init__(self, inserted_id):
                self.acknowledged = True
                self.inserted_id = inserted_id
        return InsertOneResult(doc["_id"])

    async def insert_many(self, docs):
        data = _load_data()
        for doc in docs:
            doc = deepcopy(doc)
            if "_id" not in doc:
                doc["_id"] = ObjectId()
            data[self._name].append(doc)
        _persist_data(data)

        class InsertManyResult:
            def __init__(self, count):
                self.acknowledged = True
                self.inserted_count = count
        return InsertManyResult(len(docs))

    async def update_one(self, filter_dict, update):
        data = _load_data()
        for doc in data[self._name]:
            if _matches_filter(doc, filter_dict):
                if "$set" in update:
                    for k, v in update["$set"].items():
                        _set_path_value(doc, k, v)
                if "$push" in update:
                    push_data = update["$push"]
                    for k, v in push_data.items():
                        arr = _get_path_value(doc, k) or []
                        if "$each" in v:
                            arr.extend(v["$each"])
                        else:
                            arr.append(v)
                        _set_path_value(doc, k, arr)
                _persist_data(data)

                class UpdateResult:
                    def __init__(self):
                        self.acknowledged = True
                        self.matched_count = 1
                        self.modified_count = 1
                return UpdateResult()

        class UpdateResult:
            def __init__(self):
                self.acknowledged = True
                self.matched_count = 0
                self.modified_count = 0
        return UpdateResult()

    async def delete_one(self, filter_dict):
        data = _load_data()
        for i, doc in enumerate(data[self._name]):
            if _matches_filter(doc, filter_dict):
                data[self._name].pop(i)
                _persist_data(data)

                class DeleteResult:
                    def __init__(self):
                        self.acknowledged = True
                        self.deleted_count = 1
                return DeleteResult()

        class DeleteResult:
            def __init__(self):
                self.acknowledged = True
                self.deleted_count = 0
        return DeleteResult()

    async def delete_many(self, filter_dict):
        data = _load_data()
        before = len(data[self._name])
        data[self._name] = [d for d in data[self._name] if not _matches_filter(d, filter_dict)]
        deleted = before - len(data[self._name])
        if deleted > 0:
            _persist_data(data)

        class DeleteResult:
            def __init__(self):
                self.acknowledged = True
                self.deleted_count = deleted
        return DeleteResult()

    # --- 聚合 ---
    async def count_documents(self, filter_dict=None):
        data = _load_data()
        return sum(1 for d in data[self._name] if _matches_filter(d, filter_dict))

    async def distinct(self, field):
        data = _load_data()
        values = set()
        for doc in data[self._name]:
            val = _get_path_value(doc, field)
            if val is not None:
                values.add(str(val))
        return list(values)

    def aggregate(self, pipeline):
        """支持 $vectorSearch 聚合管道。"""
        data = _load_data()

        vector_search = None
        project = None
        for stage in pipeline:
            if "$vectorSearch" in stage:
                vector_search = stage["$vectorSearch"]
            if "$project" in stage:
                project = stage["$project"]

        if not vector_search:
            return LocalCursor([])

        query_vector = vector_search.get("queryVector", [])
        limit = vector_search.get("limit", 5)

        results = []
        for doc in data[self._name]:
            if not isinstance(doc.get("text"), str):
                continue
            emb = doc.get("embedding")
            if not isinstance(emb, list) or not emb:
                continue

            score = _cosine_similarity(query_vector, emb)
            # 归一化到 [0, 1]
            normalized_score = (max(-1.0, min(1.0, score)) + 1) / 2
            doc_copy = deepcopy(doc)
            doc_copy["score"] = normalized_score
            results.append(doc_copy)

        results.sort(key=lambda d: d["score"], reverse=True)
        results = results[:limit]

        if project:
            results = [_apply_projection(doc, project) for doc in results]

        return LocalCursor(results)


# ---- 预创建实例 ----
_kb_chunks = LocalKbCollection("kb_chunks")
_kb_files = LocalKbCollection("kb_files")


def get_local_kb_collection(name):
    """获取本地存储集合。"""
    if name == "kb_chunks":
        return _kb_chunks
    elif name == "kb_files":
        return _kb_files
    raise ValueError(f"Unknown collection: {name}")


def is_local_kb_collection(collection):
    """检查是否为本地存储集合。"""
    return getattr(collection, "isLocalKbCollection", False)
